### PR TITLE
Added new `Lint/TopLevelReturnWithArgument` cop

### LINKED ISSUE #7425: New cop: Detect return with argument on the top-level
## Is your feature request related to a problem? Please describe.

`return` has been allowed on the top-level since Ruby 2.4.

```ruby
foo

return # Since Ruby 2.4.

bar # It is never executed.
```

But the argument of `return` is ignored always.

```ruby
return 1 # 1 is always ignored.
```

It will be detected since Ruby 2.7. https://github.com/ruby/ruby/commit/aa97410b0a85cb4ceb956ab943b5eee92a128411
But RuboCop doesn't.

## Describe the solution you'd like

Add a new cop to detect the problem. Maybe it will be named `Lint/TopLevelReturnWithArgument`.

## Describe alternatives you've considered

nothing

## Additional context

nothing

### PR DESCRIPTION
**Pull request by IamRaviTejaG**
Fixes #7425 

This cop detects top level return statements with argument(s).

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/lint/top_level_return_with_argument.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      #
+      # @example
+      #
+      #   # Detected since Ruby 2.7
+      #   return 1 # 1 is always ignored.
+      class TopLevelReturnWithArgument < Cop
+        MSG = 'Top level return with argument detected.'
+
+        def on_return(return_node)
+          add_offense(return_node) if ancestors_valid?(return_node) && !return_node.arguments.empty?
```
Comment (marcandre) [inline lib/rubocop/cop/lint/top_level_return_with_argument.rb:46]: `return_node.arguments.empty?` can also be written `return_node.arguments?`

File: lib/rubocop/cop/lint/top_level_return_with_argument.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        end
+
+        private
+
+        def ancestors_valid?(return_node)
+          return true if return_node.parent.nil?
+
+          if return_node.parent.instance_of?(AST::Node) && return_node.parent.parent.nil?
```
Comment (marcandre) [inline lib/rubocop/cop/lint/top_level_return_with_argument.rb:54]: Won't `return_node.parent.instance_of?(AST::Node)` always be true?

File: lib/rubocop/cop/lint/top_level_return_with_argument.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            return true
+          end
+
+          false
```
Comment (marcandre) [inline lib/rubocop/cop/lint/top_level_return_with_argument.rb:58]: I'm surprised there's no cop against `if condition; return true; end; return false` => `return condition`

File: lib/rubocop/cop/lint/top_level_return_with_argument.rb (hunk 2 above)
Comment (IamRaviTejaG) [reply lib/rubocop/cop/lint/top_level_return_with_argument.rb:54]: `return_node.parent.is_a?(AST::Node)` will always return true, but `return_node.parent.instance_of?(AST::Node)` would be true only if that object is strictly of that exact class, and not the subclass.

File: lib/rubocop/cop/lint/top_level_return_with_argument.rb (hunk 3 above)
Comment (IamRaviTejaG) [reply lib/rubocop/cop/lint/top_level_return_with_argument.rb:58]: I did write it in one line, but then it exceeded the 100 characters limit, so had to change it to this.

Comment (IamRaviTejaG) [reply lib/rubocop/cop/lint/top_level_return_with_argument.rb:58]: @marcandre There actually is a cop that suggests this, one suggests for single line if and the other suggests to remove redundant false.

File: lib/rubocop/cop/lint/top_level_return_with_argument.rb (hunk 2 above)
Comment (marcandre) [reply lib/rubocop/cop/lint/top_level_return_with_argument.rb:54]: Oh, right, my bad. Then I would say you need to tweak this test; you should not rely on the class of `Node`, only on their `type`. Here anyways I believe it is not the best anyways; you're excluding `if` for example.

File: lib/rubocop/cop/lint/top_level_return_with_argument.rb (hunk 3 above)
Comment (marcandre) [reply lib/rubocop/cop/lint/top_level_return_with_argument.rb:58]: In this case here, you can remove the `if`, the `return true\n  end\n  false` completely...

File: lib/rubocop/cop/lint/top_level_return_with_argument.rb (hunk 2 above)
Comment (IamRaviTejaG) [reply lib/rubocop/cop/lint/top_level_return_with_argument.rb:54]: But to find out the type in Ruby, we do `is_a?(ClassName)` or `instance_of?(ClassName)`. If not, then I guess I quite did not get what you suggested.

File: lib/rubocop/cop/lint/top_level_return_with_argument.rb (hunk 3 above)
Comment (IamRaviTejaG) [reply lib/rubocop/cop/lint/top_level_return_with_argument.rb:58]: The if condition checks for the types and returns things appropriately, if we remove that, the cop will report false positives.

File: lib/rubocop/cop/lint/top_level_return_with_argument.rb (hunk 1 above)
Comment (IamRaviTejaG) [reply lib/rubocop/cop/lint/top_level_return_with_argument.rb:46]: Aah yes!

File: lib/rubocop/cop/lint/top_level_return_with_argument.rb (hunk 3 above)
Comment (IamRaviTejaG) [reply lib/rubocop/cop/lint/top_level_return_with_argument.rb:58]: Updated this! Thanks for the idea. 😄

**Comment by IamRaviTejaG** [comment]
@marcandre Updated the PR. Now `return 42 if a == b` are being caught.

**Comment by marcandre** [comment]
> @marcandre Updated the PR. Now `return 42 if a == b` are being caught.

RIght, good, but it won't catch `if a == b; warn 'hey'; return 42; end`...

I was trying to hint at a more general logic. Unless I'm missing something, you can check all ancestors for either a `block`, `def` or `defs`; if you don't find any then you are still at the top level scope.

**Comment by IamRaviTejaG** [comment]
> 
> 
> > @marcandre Updated the PR. Now `return 42 if a == b` are being caught.
> 
> RIght, good, but it won't catch `if a == b; warn 'hey'; return 42; end`...
> 
> I was trying to hint at a more general logic. Unless I'm missing something, you can check all ancestors for either a `block`, `def` or `defs`; if you don't find any then you are still at the top level scope.

Now this checks for the ancestors to not belong to the `AST::BlockNode` or `AST::DefNode` class.

**Comment by marcandre** [comment]
> Now this checks for the ancestors to not belong to the `AST::BlockNode` or `AST::DefNode` class.

Much better. Please do not rely on classes ever, as this is subject to change. You should rely on the result of `Node#type` or `Node#block_type? / def_type? / defs_type?` In this case, `each_ancestor` already accepts a list of types...

**Comment by IamRaviTejaG** [comment]
> 
> 
> > Now this checks for the ancestors to not belong to the `AST::BlockNode` or `AST::DefNode` class.
> 
> Much better. Please do not rely on classes ever, as this is subject to change. You should rely on the result of `Node#type` or `Node#block_type? / def_type? / defs_type?` In this case, `each_ancestor` already accepts a list of types...

Aah, this suggestion and the resulting changes made the code so compact! Can't thank you enough for this.

File: lib/rubocop/cop/lint/top_level_return_with_argument.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+        def on_return(return_node)
+          add_offense(return_node) if ancestors_valid?(return_node) && return_node.arguments?
+        end
+
+        private
+
+        def ancestors_valid?(return_node)
+          prohibited_ancestors = return_node.each_ancestor(:block, :def, :defs)
+          return false if prohibited_ancestors.any?
```
Comment (marcandre) [inline lib/rubocop/cop/lint/top_level_return_with_argument.rb:29]: Personal preference, I'd simply write `prohibited_ancestors.none?`, no return, no if...

File: lib/rubocop/cop/lint/top_level_return_with_argument.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+      #
+      # @example
+      #
+      #   # Detected since Ruby 2.7
+      #   return 1 # 1 is always ignored.
+      class TopLevelReturnWithArgument < Cop
+        MSG = 'Top level return with argument detected.'
+
+        def on_return(return_node)
+          add_offense(return_node) if ancestors_valid?(return_node) && return_node.arguments?
```
Comment (marcandre) [inline lib/rubocop/cop/lint/top_level_return_with_argument.rb:22]: I'd switch the conditions' order, as `arguments?` is very quick while `ancestors_valid?` might be slower

**Comment by marcandre** [comment]
> Aah, this suggestion and the resulting changes made the code so compact! Can't thank you enough for this.

Cool, I'm happy you're happy about the resulting code. I have two last remarks for styling/performance...

File: lib/rubocop/cop/lint/top_level_return_with_argument.rb (hunk 5 above)
Comment (IamRaviTejaG) [reply lib/rubocop/cop/lint/top_level_return_with_argument.rb:22]: Yes, this also makes a lot of sense given Ruby's short circuit evaluation.

**Comment by IamRaviTejaG** [comment]
> Cool, I'm happy you're happy about the resulting code. I have two last remarks for styling/performance...
> 
> Also you'll have to rebase your branch

Made performance & styling related changes, and also rebased my branch. 😄

File: docs/modules/ROOT/pages/cops_lint.adoc [hunk 6]
```diff
@@ -3358,6 +3358,34 @@ def to_json(*_args)
...
+| -
+|===
+
+This cop checks for top level return with arguments. If there is a
+top-level return statement with an argument, then the argument is
+always ignored. This is detected automatically since Ruby 2.7.
+
+This cop works by validating the ancestors of the return node. A
+top-level return node's ancestors will never belong to `AST::BlockNode`
+or `AST::DefNode` class.
```
Comment (marcandre) [inline docs/modules/ROOT/pages/cops_lint.adoc:3379]: I'd remove this comment; implementation detail is too technical and not useful to cop users.

File: spec/rubocop/cop/lint/top_level_return_with_argument_spec.rb [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+  end
+
+  context 'Code segment with block level returns other than the top-level return' do
+    it 'Expects no offense from the return without arguments' do
+      expect_no_offenses(<<~RUBY)
+        foo
+
+        [1, 2, 3, 4, 5].each { |n| return n }
+
+        return # Should raise a `top level return with argument detected` offense
```
Comment (marcandre) [inline spec/rubocop/cop/lint/top_level_return_with_argument_spec.rb:39]: Change comment...

File: spec/rubocop/cop/lint/top_level_return_with_argument_spec.rb [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+  context 'Code segment with method-level return statements' do
+    it 'Expects no offense from the method-level return statement' do
+      expect_no_offenses(<<~RUBY)
+        def method
+          return 'Hello World'
+        end
+      RUBY
+    end
+
```
Comment (marcandre) [inline spec/rubocop/cop/lint/top_level_return_with_argument_spec.rb:67]: Remove this spec, it is redundant with the next spec.

File: spec/rubocop/cop/lint/top_level_return_with_argument_spec.rb [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        bar
+      RUBY
+    end
+
+    it 'Expects an offense from the return with arguments' do
+      expect_offense(<<~RUBY)
+        if a == b; warn 'hey'; return 42; end
+                               ^^^^^^^^^ Top level return with argument detected.
+      RUBY
+    end
```
Comment (marcandre) [inline spec/rubocop/cop/lint/top_level_return_with_argument_spec.rb:130]: Seems redundant with previous spec

File: spec/rubocop/cop/lint/top_level_return_with_argument_spec.rb [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+    it 'Expects no offense from the return without arguments' do
+      expect_no_offenses(<<~RUBY)
+        if a == b; warn 'hey'; return; end
+      RUBY
+    end
+
```
Comment (marcandre) [inline spec/rubocop/cop/lint/top_level_return_with_argument_spec.rb:137]: Seems redundant with next spec

**Comment by marcandre** [comment]
Implementation is perfect now 🎉. I think many specs can be removed and some doc can be modified.

File: lib/rubocop/cop/lint/top_level_return_with_argument.rb [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  module Cop
+    module Lint
+      # This cop checks for top level return with arguments. If there is a
+      # top-level return statement with an argument, then the argument is
+      # always ignored. This is detected automatically since Ruby 2.7.
+      #
+      # This cop works by validating the ancestors of the return node. A
+      # top-level return node's ancestors will never belong to `AST::BlockNode`
+      # or `AST::DefNode` class.
+      #
```
Comment (marcandre) [inline lib/rubocop/cop/lint/top_level_return_with_argument.rb:13]: Please update and put inside the class so it doesn't show in the doc

Comment (IamRaviTejaG) [reply lib/rubocop/cop/lint/top_level_return_with_argument.rb]: Updated this and added it inside class.

File: spec/rubocop/cop/lint/top_level_return_with_argument_spec.rb (hunk 10 above)
Comment (IamRaviTejaG) [reply spec/rubocop/cop/lint/top_level_return_with_argument_spec.rb]: Yes, removed this.

File: spec/rubocop/cop/lint/top_level_return_with_argument_spec.rb (hunk 9 above)
Comment (IamRaviTejaG) [reply spec/rubocop/cop/lint/top_level_return_with_argument_spec.rb]: Yes, removed this spec.

File: spec/rubocop/cop/lint/top_level_return_with_argument_spec.rb (hunk 8 above)
Comment (IamRaviTejaG) [reply spec/rubocop/cop/lint/top_level_return_with_argument_spec.rb]: Done.

File: spec/rubocop/cop/lint/top_level_return_with_argument_spec.rb (hunk 7 above)
Comment (IamRaviTejaG) [reply spec/rubocop/cop/lint/top_level_return_with_argument_spec.rb]: Removed.

File: docs/modules/ROOT/pages/cops_lint.adoc (hunk 6 above)
Comment (IamRaviTejaG) [reply docs/modules/ROOT/pages/cops_lint.adoc]: Makes sense to me, removed this. Added this inside the class instead.

**Comment by IamRaviTejaG** [comment]
Hey @bbatsov, can you please review this PR. 😄

**Comment by marcandre** [comment]
It's ok, I probably should have simply merged it, looks really good. Looks like there are conflicts again, could you rebase and I'll merge it?

**Comment by IamRaviTejaG** [comment]
@marcandre One test failed erroneously, and I cannot seem to re-run it, can you please re-run it? Once that passes this should be good to merge.

**Comment by marcandre** [comment]
Thanks @IamRaviTejaG 🎉

**Pull request merged**